File: src/main/java/me/testaccount666/serversystem/userdata/listener/UserJoinListener.java
package me.testaccount666.serversystem.userdata.listener;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.userdata.UserManager;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;

public class UserJoinListener implements Listener {

    @EventHandler(priority = EventPriority.LOWEST)
    public void onUserJoin(PlayerJoinEvent event) {
        var cachedUserOptional = ServerSystem.Instance.getRegistry().getService(UserManager.class).getUser(event.getPlayer().getUniqueId());

        if (cachedUserOptional.isEmpty()) throw new RuntimeException("Couldn't cache User '${event.getPlayer().getName()}'! This should not happen!");

        var cachedUser = cachedUserOptional.get();

        if (cachedUser.isOnlineUser()) return;

        cachedUser.convertToOnlineUser();

        Bukkit.getScheduler().runTaskLater(ServerSystem.Instance, () -> {
            var user = cachedUser.getOfflineUser();
            if (user.getLogoutPosition() == null) return;

            event.getPlayer().teleport(user.getLogoutPosition());
        }, 10L);
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/userdata/listener/UserJoinListener.kt
package me.testaccount666.serversystem.userdata.listener

import me.testaccount666.paperktx.scheduler.skedule.okkero.schedule
import me.testaccount666.serversystem.ServerSystem
import me.testaccount666.serversystem.extensions.getService
import me.testaccount666.serversystem.userdata.UserManager
import org.bukkit.event.*
import org.bukkit.event.player.PlayerJoinEvent

class UserJoinListener : Listener {
    @EventHandler(priority = EventPriority.LOWEST)
    fun onUserJoin(event: PlayerJoinEvent) {
        val cachedUser = getService<UserManager>().getUserOrNull(event.getPlayer().uniqueId) ?: run {
            throw RuntimeException("Couldn't cache User '${event.getPlayer().name}'! This should not happen!")
        }
        if (cachedUser.isOnlineUser) return

        cachedUser.convertToOnlineUser()

        ServerSystem.instance.schedule {
            waitFor(10L)

            val user = cachedUser.offlineUser.takeIf { it.logoutPosition != null } ?: return@schedule
            event.getPlayer().teleport(user.logoutPosition!!)
        }
    }
}